flink-runtime/src/main/java/org/apache/flink/runtime/taskmanager/Task.java

```diff
@@ -637,6 +636,11 @@ else if (current == ExecutionState.CANCELING) {
 
 			TaskKvStateRegistry kvStateRegistry = kvStateService.createKvStateTaskRegistry(jobId, getJobVertexId());
 
+			InputGate[] gates = new InputGate[inputGates.length];
+			for (int i = 0; i < inputGates.length; i++) {
```

Comment:
To be honest, I liked wrapping in constructor more. This way we keep in `Task` and `RuntimeEnvironment` same objects and there is less risk for them to diverge and forget that we call real gate and not wrapper. What do you think?

I would at least move the wrapping loop to a separate static method.

Reply:
Yes, I also did it in the constructor in first version. Later I think it might be better to make `InputGateWithMetrics` involved in less logics such as `close`, so I change it as now to keep the previous arrays of `SingleInputGate` in task.

I think it also makes sense to do it in constructor, then I would fix it.

Reply:
Forgot another concern for not doing in the constructor. It would make maintain two arrays of `SingleInputGate` and `InputGate` separately. Because we still need to return the array of `SingleInputGate` for `NetworkEnvironment#setup` currently. What do you think?